c/tskit/trees.c

```diff
@@ -1858,6 +1859,139 @@ tsk_treeseq_diversity(tsk_treeseq_t *self,
         diversity_summary_func, num_windows, windows, result, options);
 }
 
+static int
+inheritance_covariance_summary_func(size_t state_dim, double *state, size_t TSK_UNUSED(result_dim),
+        double *result, void *TSK_UNUSED(params))
+{
+    const double *x = state;
+    size_t j;
+
+    for (j = 0; j < state_dim; j++) {
+        result[j] = (x[j] * x[j]) / 2;
+    }
+    return 0;
+}
+
+int
+tsk_treeseq_inheritance_covariance(tsk_treeseq_t *self,
+        tsk_size_t num_weights, double *weights,
+        tsk_size_t num_windows, double *windows, double *result, tsk_flags_t options)
+{
+    tsk_size_t num_samples = self->num_samples;
+    size_t j, k;
+    int ret;
+    double *row, *new_row;
+    double *means = calloc(num_weights, sizeof(double));
+    double *new_weights = malloc((num_weights + 1) * num_samples * sizeof(double));
+
+    if (new_weights == NULL || means == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // center weights
+    for (j = 0; j < num_samples; j++) {
+        row = GET_2D_ROW(weights, num_weights, j);
+        for (k = 0; k < num_weights; k++) {
+            means[k] += row[k];
+        }
+    }
+    for (k = 0; k < num_weights; k++) {
+        means[k] /= (float) num_samples;
+    }
+    for (j = 0; j < num_samples; j++) {
+        row = GET_2D_ROW(weights, num_weights, j);
+        new_row = GET_2D_ROW(new_weights, num_weights, j);
+        for (k = 0; k < num_weights; k++) {
+            new_row[k] = row[k] - means[k];
+        }
+    }
+
+    ret = tsk_treeseq_general_stat(self,
+            num_weights, new_weights, num_weights,
+            inheritance_covariance_summary_func, NULL, num_windows, windows, result, options);
+
+out:
+    tsk_safe_free(means);
+    tsk_safe_free(new_weights);
+    return ret;
+}
+
+static int
+inheritance_correlation_summary_func(size_t state_dim, double *state, size_t TSK_UNUSED(result_dim),
+        double *result, void *TSK_UNUSED(params))
+{
+    const double *x = state;
+    double p;
+    size_t j;
+
+    p = x[state_dim - 1];
+    for (j = 0; j < state_dim - 1; j++) {
```

Comment:
Is there any way p could be negative? Do we need to guard against this? Numpy returns a NaN if -1 is passed to sqrt, so many it's OK to just let it happen like we're doing with division by zero.

Reply:
Stupid Jerome! I just read the first bit of the line again and realised why it was doing it. Maybe clearer to have an if statement for this?